### PR TITLE
Device disconnected message

### PR DESCRIPTION
**Pull request by LLFourn**
Builds on #80.

The purpose of the PR is to notify the coordinator when a downstream device physically disconnects.

So far I've done two refactorings before this:

- All messages sent from a device include a `from` field
- Simplify message sending by improving how we use the `Direction` type parameter to the messages

Depending on how things go it might be worth reviewing those two changes and merging them first.

### CODE REVIEW

**Comment by musdom** [comment]
Tested frostsnapp. Mistakenly tested branch devices with master frostsnapp and found bug.

**Comment by LLFourn** [comment]
> Tested frostsnapp. Mistakenly tested branch devices with master frostsnapp and found bug.

So I looked for how I could have changed the protocol in this branch. I couldn't find anything so I guess we'll look the other way on this. As long as it all works on this PR :P

File: device/src/esp32_run.rs [hunk 1]
```diff
@@ -159,10 +143,7 @@ where
                 if downstream_serial.find_and_remove_magic_bytes() {
                     downstream_active = true;
                     ui.set_downstream_connection_state(true);
-                    sends_upstream.push(DeviceSendMessage::Debug {
-                        message: "Device read magic bytes from another device!".to_string(),
-                        device: frost_signer.clone().device_id(),
-                    });
+                    sends_upstream.debug("Device read magic bytes from another device!");
```
Comment (nickfarrow) [inline device/src/esp32_run.rs:146]: Could we maybe change .`debug` to `.send_debug`?

**Comment by nickfarrow** [comment]
Nice I like what has been done with ` Direction`.

Everything looks good to me so far and the fix for duplicate device showing in app works.

Where in the coordinator code will we model connections so we can disconnect multiple downstream devices from a single `Disconnected` message?

File: device/src/esp32_run.rs (hunk 1 above)
Comment (LLFourn) [reply device/src/esp32_run.rs:146]: I don't mind either way so I did it.

**Comment by LLFourn** [comment]
Somewhere is `usb_serial_manager.rs`. I think the harder problem is informing the coordinator about which device is connected to which.

**Comment by LLFourn** [comment]
I've done the device disconnected message as well. It's working well on linux app but weird stuff is happening on android.

**Comment by musdom** [comment]
Works on both android and linux. Is the menu bar gone?

**Comment by LLFourn** [comment]
> Works on both android and linux. Is the menu bar gone?

Yes it was not doing anything so I thought to simplify. Want it back?

**Comment by musdom** [comment]
Ok no need, all good

File: device/src/esp32_run.rs [hunk 2]
```diff
@@ -61,15 +66,15 @@ where
                         signer: frost_signer.clone(),
                     })
                     .unwrap();
-                ui.misc_print("New secret generated and saved");
+                ui.dispaly_debug("New secret generated and saved");
```
Comment (nickfarrow) [inline device/src/esp32_run.rs:69]: typo
dispaly_debug -> display_debug

File: frostsnap_coordinator/src/usb_serial_manager.rs [hunk 3]
```diff
@@ -273,6 +273,24 @@ impl UsbSerialManager {
...
+                    DeviceSendBody::DisconnectDownstream => {
+                        if let Some(device_list) = self.reverse_device_ports.get_mut(&serial_number)
+                        {
+                            if let Some((i, _)) = device_list
+                                .iter()
+                                .enumerate()
+                                .find(|(_, device_id)| **device_id == message.from)
+                            {
+                                let index_of_disconnection = i + 1;
+                                while device_list.len() > index_of_disconnection {
```
Comment (nickfarrow) [inline frostsnap_coordinator/src/usb_serial_manager.rs:285]: Nice LGTM. Confirmed working on multiple ports

**Review by nickfarrow — APPROVED** [review]

**Pull request merged**